client/components/migration/migration-runner.js

```diff
@@ -0,0 +1,251 @@
+ */
+const migrationStateTransitions = {
+    stateInit: {
+        success: 'stateInstalling',
+        fail: 'stateInit',
+        callback: fetchAPICall(markMigrationStartedAPICall(MIGRATION_STATE_ENUM.STARTED)),
+    },
+    stateErrorInit: {
+        success: 'stateInit',
+        fail: 'stateErrorInit',
+        callback: stateErrorHandlingAPICall(MIGRATION_STATE_ENUM.ERROR_STARTED),
+    },
+    stateInstalling: {
+        success: 'stateActivating',
+        fail: 'stateErrorInstalling',
+        callback: fetchAPICall(installPluginAPICall),
+    },
+    stateErrorInstalling: {
+        success: 'stateInstalling',
+        fail: 'stateErrorInstalling',
+        callback: stateErrorHandlingAPICall(MIGRATION_STATE_ENUM.ERROR_INSTALLING),
+    },
+    stateActivating: {
+        success: 'stateDeactivating', // TODO: This should be stateDBMigrating to migrate DB.
+        fail: 'stateErrorActivating',
+        callback: fetchAPICall(activatePluginAPICall),
+    },
+    stateErrorActivating: {
+        success: 'stateActivating',
+        fail: 'stateErrorActivating',
+        callback: stateErrorHandlingAPICall(MIGRATION_STATE_ENUM.ERROR_ACTIVATING),
+    },
+    stateDB: {
+        success: 'stateDeactivating',
+        fail: 'stateErrorDB',
+        callback: new Promise((resolve) => resolve({status: 200})), // TODO: DB migration
+    },
+    stateErrorDB: {
+        success: 'stateDBMigrating',
+        fail: 'stateErrorDB',
+        callback: stateErrorHandlingAPICall(MIGRATION_STATE_ENUM.ERROR_DB_MIGRATION),
+    },
+    stateDeactivating: {
+        success: 'stateDone',
+        fail: 'stateErrorDeactivating',
+        callback: fetchAPICall(deactivateWCSTPluginAPICall),
+    },
+    stateErrorDeactivating: {
+        success: 'stateDeactivating',
+        fail: 'stateErrorDeactivating',
+        callback: stateErrorHandlingAPICall(MIGRATION_STATE_ENUM.ERROR_DEACTIVATING),
+    },
+    stateDone: { // Done state.
+        success: null,
+        fail: null,
+        callback: null
+    }
+}
+
+// Any of the state that halts the machine. Including "done" or any errors.
+const stopStates = ['stateDone', 'stateErrorInstalling', 'stateErrorActivating', 'stateErrorDB', 'stateErrorDeactivating'];
+
+/**
+ * This function runs the entire migration based on what the previousMigrationState is. It will stop
+ * when it lands on an "error state" or when the migration is completed.
+ *
+ * @param {number} previousMigrationState This is the "wcshipping_migration_state" stored in the option table. This is the last state recorded for this migration before it stopped.
+ */
+const installAndActivatePlugins = async(previousMigrationState) => {
+    const runNext = async (migrationState) => {
+        if (!migrationState) {
+            // Nothing to run, do nothing and return.
+            return;
+        }
+
+        const currentMigrationState = migrationStateTransitions[migrationState];
```

Comment:
should we check if `migrationState` is included in `migrationStateTransitions`?

Reply:
Thanks! I added a function `getMigrationStateByName(stateName)` in 504f9c5f. It will now check if the state exists before using it. It will throw exception otherwise. Let me know what you think!